### DISCUSSION TITLE
Custom Board Code on LyraT

### DISCUSSION THREAD

**Discussion by lufa07** [pschatzmann/arduino-audio-driver#13] [Q&A] [answered]
1. The selected scenario/sketch: audiotools-custom-max.ino
2. what exactly I am trying to do: Trying to make this code work on standard LyraT board. I am doing this as part of I2S FW bring up for my custom design board. If this works, I will modify the pins according to my design and try to reproduce it on the custom PCB
3. Hardware: Currently working on LyraT but later on move to custom designed board
4. Software versions: Arduino IDE 2.3.0, audio drivers updated today
5. Problem: audiotools-custom-max.ino sketch on LyraT with its pin definitions explicitly defined to mimick the FW to be later on ported to a custom board.

Please let me know if audiotools-custom-max.ino sketch could be mimicked to work with LyraT

**Comment by pschatzmann** [comment]
Sure, this is very flexible, just adjust the pin definitions. By the way the following Lyrat boards are already predefined 
- LyratV43{AudioDriverES8388, PinsLyrat43};
- LyratV42{AudioDriverES8388, PinsLyrat42};

e.g.PinsLyrat42Class is 

```C++

class PinsLyrat42Class : public DriverPins {
 public:
  PinsLyrat42Class() {
    // sd pins
    addSPI(ESP32PinsSD);
    // add i2c codec pins: scl, sda, port, frequency
    addI2C(PinFunction::CODEC, 23, 18);
    // add i2s pins: mclk, bck, ws,data_out, data_in ,(port)
    addI2S(PinFunction::CODEC, 0, 5, 25, 26, 35);

    // add other pins
    addPin(PinFunction::KEY, 36, PinLogic::InputActiveLow, 1);
    addPin(PinFunction::KEY, 39, PinLogic::InputActiveLow, 2);
    addPin(PinFunction::KEY, 33, PinLogic::InputActiveLow, 3);
    addPin(PinFunction::KEY, 32, PinLogic::InputActiveLow, 4);
    addPin(PinFunction::KEY, 13, PinLogic::InputActiveLow, 5);
    addPin(PinFunction::KEY, 27, PinLogic::InputActiveLow, 6);
    addPin(PinFunction::AUXIN_DETECT, 12, PinLogic::InputActiveLow);
    addPin(PinFunction::PA, 21, PinLogic::Output);
    addPin(PinFunction::LED, 22, PinLogic::Output, 1);
    addPin(PinFunction::LED, 19, PinLogic::Output, 2);
  }
};
```

**Reply by lufa07** [reply]
Pinouts defined as:
// I2C
#define SDAPIN               18 // I2C Data,  Adafruit ESP32 S3 3, Sparkfun Thing Plus C 23
#define SCLPIN               23 // I2C Clock, Adafruit ESP32 S3 4, Sparkfun Thing Plus C 22
#define I2CSPEED        100000 // Clock Rate
#define ES8388ADDR        0x20 // Address of ES8388 I2C port

// I2S, your configuration for the ES8388 board
#define MCLKPIN             0 // Master Clock
#define BCLKPIN             5 // Bit Clock
#define WSPIN                25 // Word select
#define DOPIN               26 // This is connected to DI on ES8388 (MISO)
#define DIPIN               35 // This is connected to DO on ES8388 (MOSI)

Since the example did not work as expected, I made some additional changes as below:
  auto i2s_config = i2s_out_stream.defaultConfig(RXTX_MODE);
  i2s_config.input_device = ADC_INPUT_ALL;
  i2s_config.output_device = DAC_OUTPUT_LINE2;
  i2s_config.sd_active = false;
  //i2s_config.i2s.fmt = I2S_NORMAL;
  //i2s_config.i2s.mode = MODE_SLAVE;
  i2s_config.copyFrom(audio_info);  
  i2s_in_stream.begin(i2s_config); // this should apply I2C and I2S configuration
  i2s_out_stream.begin(i2s_config); // this should apply I2C and I2S configuration

  // Setup sine wave
  Serial.println("Sine wave begin...");
  sine_wave.begin(audio_info, N_C6); // 493.88 Hz

  Serial.println("Setup completed ...");
  
  Even then I do not hear anything on my earphone connected to LyraT

**Reply by pschatzmann** [reply]
I suggest that you try first with the predefined board. Your code does not show what class the i2s_out_stream is!
You did not define the I2S pins anywhere!

**Reply by lufa07** [reply]
Adding full code here for clarity:

```
/**
 * @brief We define a custom board with the i2c and i2s pins and output a sine
 * with the help of the AudioTools I2SCodecStream
 * @author phil schatzmann
 */

#include "AudioTools.h" // install https://github.com/pschatzmann/arduino-audio-tools
#include "AudioLibs/I2SCodecStream.h"
#include "AudioBoard.h"

// I2C
#define SDAPIN               18 // I2C Data,  Adafruit ESP32 S3 3, Sparkfun Thing Plus C 23
#define SCLPIN               23 // I2C Clock, Adafruit ESP32 S3 4, Sparkfun Thing Plus C 22
#define I2CSPEED        100000 // Clock Rate
#define ES8388ADDR        0x20 // Address of ES8388 I2C port

// I2S, your configuration for the ES8388 board
#define MCLKPIN             0 // Master Clock
#define BCLKPIN             5 // Bit Clock
#define WSPIN                25 // Word select
#define DOPIN               26 // This is connected to DI on ES8388 (MISO)
#define DIPIN               35 // This is connected to DO on ES8388 (MOSI)

AudioInfo                     audio_info(44200, 1, 16);                // sampling rate, # channels, bit depth
SineWaveGenerator<int16_t>    sine_wave(32767);                        // amplitude
GeneratedSoundStream<int16_t> sound_stream(sine_wave);                 // sound generator
DriverPins                    my_pins;                                 // board pins
AudioBoard                    audio_board(AudioDriverES8388, my_pins); // audio board
I2SCodecStream                i2s_out_stream(audio_board);             // i2s coded
//I2SCodecStream                i2s_in_stream(audio_board);             // i2s coded
StreamCopy                    copier(i2s_out_stream, sound_stream);    // stream copy sound generator to i2s codec
TwoWire                       myWire = TwoWire(0);                     // universal I2C interface

void setup() {
  // Setup logging
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);
  LOGLEVEL_AUDIODRIVER = AudioDriverWarning;
  delay(2000);

  Serial.println("Setup starting...");

  Serial.println("I2C pin ...");
  my_pins.addI2C(PinFunction::CODEC, SCLPIN, SDAPIN, ES8388ADDR, I2CSPEED, myWire);
  Serial.println("I2S pin ...");
  my_pins.addI2S(PinFunction::CODEC, MCLKPIN, BCLKPIN, WSPIN, DOPIN, DIPIN);

  Serial.println("Pins begin ..."); 
  my_pins.begin();

  Serial.println("Board begin ..."); 
  audio_board.begin();

  Serial.println("I2S begin ..."); 
  auto i2s_config = i2s_out_stream.defaultConfig(RXTX_MODE);
  i2s_config.input_device = ADC_INPUT_ALL;
  i2s_config.output_device = DAC_OUTPUT_LINE2;
  i2s_config.sd_active = false;
  //i2s_config.i2s.fmt = I2S_NORMAL;
  //i2s_config.i2s.mode = MODE_SLAVE;
  i2s_config.copyFrom(audio_info);  
  //i2s_in_stream.begin(i2s_config); // this should apply I2C and I2S configuration
  i2s_out_stream.begin(i2s_config); // this should apply I2C and I2S configuration

  /*  // configure codec
  CodecConfig cfg;
  cfg.input_device = ADC_INPUT_ALL;
  cfg.output_device = DAC_OUTPUT_ALL;
  cfg.i2s.bits = BIT_LENGTH_16BITS;
  cfg.i2s.rate = RATE_44K;
  cfg.i2s.fmt = I2S_NORMAL;
  cfg.i2s.mode = MODE_SLAVE; 
  audio_board.begin(cfg);   */

  // Setup sine wave
  Serial.println("Sine wave begin...");
  sine_wave.begin(audio_info, N_C6); // 493.88 Hz

  Serial.println("Setup completed ...");
}

// Arduino loop - copy sound to out
void loop() { copier.copy(); }
```

I believe my pin definitions are according to PinsLyrat43

Debug Screen:
![image](https://github.com/pschatzmann/arduino-audio-driver/assets/161609737/37364ec3-034a-4fe8-90c9-3d9f64eba69f)

**Reply by pschatzmann** [reply]
very confusing, you are mixing the APIs from different levels! I2SCodecStream is already configuring the codec, so there is no need for your additional logic. I suggest that you clean up your code and then let it run with the log level info or debug, so that you can check what it is doing...

By the way if you work with the LyraT, there can only be one I2S connection (which can be douplex, so that you can read from it and write to it at the same time)!

```C++
#include "AudioTools.h" // install https://github.com/pschatzmann/arduino-audio-tools
#include "AudioLibs/I2SCodecStream.h"
#include "AudioBoard.h"

// I2C
#define SDAPIN 18 // I2C Data, Adafruit ESP32 S3 3, Sparkfun Thing Plus C 23
#define SCLPIN 23 // I2C Clock, Adafruit ESP32 S3 4, Sparkfun Thing Plus C 22
#define I2CSPEED 100000 // Clock Rate
#define ES8388ADDR 0x20 // Address of ES8388 I2C port

// I2S, your configuration for the ES8388 board
#define MCLKPIN 0 // Master Clock
#define BCLKPIN 5 // Bit Clock
#define WSPIN 25 // Word select
#define DOPIN 26 // This is connected to DI on ES8388 (MISO)
#define DIPIN 35 // This is connected to DO on ES8388 (MOSI)

AudioInfo audio_info(44200, 1, 16); // sampling rate, # channels, bit depth
SineWaveGenerator<int16_t> sine_wave(32767); // amplitude
GeneratedSoundStream<int16_t> sound_stream(sine_wave); // sound generator
DriverPins my_pins; // board pins
AudioBoard audio_board(AudioDriverES8388, my_pins); // audio board
I2SCodecStream i2s_out_stream(audio_board); // i2s coded
StreamCopy copier(i2s_out_stream, sound_stream); // stream copy sound generator to i2s codec
TwoWire myWire = TwoWire(0); // universal I2C interface

void setup() {
// Setup logging
Serial.begin(115200);
AudioLogger::instance().begin(Serial, AudioLogger::Warning);
LOGLEVEL_AUDIODRIVER = AudioDriverWarning;
delay(2000);

Serial.println("Setup starting...");

Serial.println("I2C pin ...");
my_pins.addI2C(PinFunction::CODEC, SCLPIN, SDAPIN, ES8388ADDR, I2CSPEED, myWire);
Serial.println("I2S pin ...");
my_pins.addI2S(PinFunction::CODEC, MCLKPIN, BCLKPIN, WSPIN, DOPIN, DIPIN);

Serial.println("I2S begin ...");
auto i2s_config = i2s_out_stream.defaultConfig(RXTX_MODE);
i2s_config.input_device = ADC_INPUT_ALL;
i2s_config.output_device = DAC_OUTPUT_LINE2;
i2s_config.sd_active = false;
//i2s_config.i2s.fmt = I2S_NORMAL;
//i2s_config.i2s.mode = MODE_SLAVE;
i2s_config.copyFrom(audio_info);
i2s_out_stream.begin(i2s_config); // this should apply I2C and I2S configuration

// Setup sine wave
Serial.println("Sine wave begin...");
sound_stream.begin(audio_info);
sine_wave.begin(audio_info, N_C6); // 493.88 Hz

Serial.println("Setup completed ...");
}

// Arduino loop - copy sound to out
void loop() { copier.copy(); }`
```

**Reply by lufa07** [reply]
I take it my code has been altered a lot; hence I run the audiotools_custom_max.ino and only change the pin numbers in the #defines. 

Code as below:
```
/**
 * @brief We define a custom board with the i2c and i2s pins and output a sine
 * with the help of the AudioTools I2SCodecStream
 * @author phil schatzmann
 */

#include "AudioTools.h" // install https://github.com/pschatzmann/arduino-audio-tools
#include "AudioLibs/I2SCodecStream.h"

// I2C
#define SDAPIN               18 // I2C Data,  Adafruit ESP32 S3 3, Sparkfun Thing Plus C 23
#define SCLPIN               23 // I2C Clock, Adafruit ESP32 S3 4, Sparkfun Thing Plus C 22
#define I2CSPEED        100000 // Clock Rate
#define ES8388ADDR        0x10 // Address of ES8388 I2C port

// I2S, your configuration for the ES8388 board
#define MCLKPIN             0 // Master Clock
#define BCLKPIN             5 // Bit Clock
#define WSPIN                25 // Word select
#define DOPIN               26 // This is connected to DI on ES8388 (MISO)
#define DIPIN               35 // This is connected to DO on ES8388 (MOSI)

AudioInfo                     audio_info(44200, 2, 16);                // sampling rate, # channels, bit depth
SineWaveGenerator<int16_t>    sine_wave(32000);                        // amplitude
GeneratedSoundStream<int16_t> sound_stream(sine_wave);                 // sound generator
DriverPins                    my_pins;                                 // board pins
AudioBoard                    audio_board(AudioDriverES8388, my_pins); // audio board
I2SCodecStream                i2s_out_stream(audio_board);             // i2s coded
StreamCopy                    copier(i2s_out_stream, sound_stream);    // stream copy sound generator to i2s codec
TwoWire                       myWire = TwoWire(0);                     // universal I2C interface

void setup() {
  // Setup logging
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);
  LOGLEVEL_AUDIODRIVER = AudioDriverWarning;
  delay(2000);

  Serial.println("Setup starting...");

  Serial.println("I2C pin ...");
  my_pins.addI2C(PinFunction::CODEC, SCLPIN, SDAPIN, ES8388ADDR, I2CSPEED, myWire);
  Serial.println("I2S pin ...");
  my_pins.addI2S(PinFunction::CODEC, MCLKPIN, BCLKPIN, WSPIN, DOPIN, DIPIN);

  Serial.println("Pins begin ..."); 
  my_pins.begin();

  Serial.println("Board begin ..."); 
  audio_board.begin();

  Serial.println("I2S begin ..."); 
  auto i2s_config = i2s_out_stream.defaultConfig();
  i2s_config.copyFrom(audio_info);  
  i2s_out_stream.begin(i2s_config); // this should apply I2C and I2S configuration

  // Setup sine wave
  Serial.println("Sine wave begin...");
  sine_wave.begin(audio_info, N_B4); // 493.88 Hz

  Serial.println("Setup completed ...");
}

// Arduino loop - copy sound to out
void loop() { copier.copy(); }

```

Debug as below:
![image](https://github.com/pschatzmann/arduino-audio-driver/assets/161609737/6dd3d6ae-f478-4756-bfa5-faf16371e313)


Still I do not hear anything on my earphone :(

Also, to ensure my setup works fine, I tested the output.ino from audiokit, the sketch works and I hear some tone on my earphone.

**Reply by pschatzmann** [reply]
The custom_max example is thought to be used e.g. with the standard ESP32 i2s API. If you want to use it with the AudioTools you need to that that example!

**Reply by pschatzmann** [reply]
go back to my last answer!

**Comment by lufa07** [comment]
> The custom_max example is thought to be used e.g. with the standard ESP32 i2s API. If you want to use it with the AudioTools you need to that that example!

You mean define my own AudioKit with the necessary pins?

**Reply by pschatzmann** [reply]
No, Arduino ESP32 comes with it's own I2S API that you could use instead of the AudioTools. Or you could use the IDF API.
By the way you are missing critical pins: e.g. the PA. Scroll up to see the original setup for PinsLyrat42Class...

**Reply by lufa07** [reply]
I think for bring up it only makes sense to call the LyraT43 kit & change the underlying pin definitions rather having to use the custom sketches.

Also, from LyraT schematics, I understand the PAs are used for the on-board speakers. But since I connect earphones, the PAs dont engage because the DAC out lines are disconnected.

**Reply by pschatzmann** [reply]
Did you set the log level to Info as recommended to check the reported settings ?
By the way your sketch still does not look like my proposal!

**Reply by lufa07** [reply]
Apologies, your code hadn't fully loaded when I checked it. However, I flashed your code on LyraT and please find the Debug Outputs below:
![image](https://github.com/pschatzmann/arduino-audio-driver/assets/161609737/23b7aed4-7377-4656-b3dd-86cd1a58c3f9)

Since in e8388.h, the address is defined as `#define ES8388_ADDR 0x10 ` , I changed the address from your proposal to 0x10 & the output log from Info is as below:
![image](https://github.com/pschatzmann/arduino-audio-driver/assets/161609737/8a3de111-7020-4174-9037-d352e17616f8)

In both cases of addresses of 0x10 & 0x20 of the ES8388, the output shows 0x10. I think the issue seems that the codec address doesnt seem to load properly.

I was just wondering if somebody else faced this issue or you, yourself during development & test on a standard audio board with pins explicitly defined in sketch.

Just reiterating, streams-audiokit-audiokit.ino works fine on the same LyraT

**Reply by pschatzmann** [reply]
Where did you define the address
Did you run a portscan

**Reply by lufa07** [reply]
Yes I have:
![image](https://github.com/pschatzmann/arduino-audio-driver/assets/161609737/7d72e379-070f-4227-8059-d4113b1c998f)

From your proposed code, Line  9 `#define ES8388ADDR 0x10 // Address of ES8388 I2C port`, passed on Lin 37 `my_pins.addI2C(PinFunction::CODEC, SCLPIN, SDAPIN, ES8388ADDR, I2CSPEED, myWire);`

Are you able to replicate this on your Test Audio Kit?

**Reply by pschatzmann** [reply]
Not sure with what version you are working. Initially this was just used as comment and the logic update to use it from the addI2C has just been added very recently.
Did you check in the log if the correct i2c pins are used...

**Reply by lufa07** [reply]
i updated the following code `AudioLogger::instance().begin(Serial, AudioLogger::Info);
LOGLEVEL_AUDIODRIVER = AudioDriverInfo;` and changed the "Core Debug Level = Info".

However, I do not get the print `AD_LOGI("setting up I2C scl: %d, sda: %d", scl, sda);`

Debug as below:
![image](https://github.com/pschatzmann/arduino-audio-driver/assets/161609737/95856a98-0518-4d3f-be96-ada0d95bc91f)

**Reply by pschatzmann** [reply]
I don't see how this could be possible: try LOGLEVEL_AUDIODRIVER = AudioDriverDebug

**Reply by lufa07** [reply]
I cloned your repo again now, to have the latest files.

LOGLEVEL_AUDIODRIVER = AudioDriverDebug output as below:
![image](https://github.com/pschatzmann/arduino-audio-driver/assets/161609737/fa13f9a2-bf8e-4713-a5ef-1a97f4710430)

In your proposal, `my_pins.begin();`, `audio_board.begin();`, were missed. I think thats why the Info were not getting printed.

Are you able to reproduce this on your test system?

**Reply by pschatzmann** [reply]
No, in this case you forgot to call i2s_out_stream.begin()!
This is automatically calling audio_board.begin() which calls my_pins.begin(). 
If you scroll up to my proposed code: it is definitely there...

But I am wondering: why is it still not working ? The pins are right now and the address as well...
I don't see what could be wrong. Maybe some loose connections

**Reply by lufa07** [reply]
Now I think either you can try reproducing it in your test system or I can hard code your PinsLyrat43Class and use the program from audio-kit.

I am using LyraT standard kit. No wires, ES8388 is soldered on-board.

Also, I commented out the `my_pins.begin();` & `audio_board.begin();` but `i2s_out_stream.begin(i2s_config); // this should apply I2C and I2S configuration` called and I see no Info prints :)

![image](https://github.com/pschatzmann/arduino-audio-driver/assets/161609737/a8a981a8-9b65-452a-a992-008e9e050012)

Full code FYR:
```
#include "AudioTools.h" // install https://github.com/pschatzmann/arduino-audio-tools
#include "AudioLibs/I2SCodecStream.h"
#include "AudioBoard.h"

// I2C
#define SDAPIN 18 // I2C Data, Adafruit ESP32 S3 3, Sparkfun Thing Plus C 23
#define SCLPIN 23 // I2C Clock, Adafruit ESP32 S3 4, Sparkfun Thing Plus C 22
#define I2CSPEED 100000 // Clock Rate
#define ES8388ADDR 0x10 // Address of ES8388 I2C port

// I2S, your configuration for the ES8388 board
#define MCLKPIN 0 // Master Clock
#define BCLKPIN 5 // Bit Clock
#define WSPIN 25 // Word select
#define DOPIN 26 // This is connected to DI on ES8388 (MISO)
#define DIPIN 35 // This is connected to DO on ES8388 (MOSI)

AudioInfo audio_info(44200, 1, 16); // sampling rate, # channels, bit depth
SineWaveGenerator<int16_t> sine_wave(32767); // amplitude
GeneratedSoundStream<int16_t> sound_stream(sine_wave); // sound generator
DriverPins my_pins;   // board pins
AudioBoard audio_board(AudioDriverES8388, my_pins); // audio board
I2SCodecStream i2s_out_stream(audio_board); // i2s coded
StreamCopy copier(i2s_out_stream, sound_stream); // stream copy sound generator to i2s codec
TwoWire myWire = TwoWire(0); // universal I2C interface

void setup() {
// Setup logging
Serial.begin(115200);
AudioLogger::instance().begin(Serial, AudioLogger::Info);
LOGLEVEL_AUDIODRIVER = AudioDriverDebug;
delay(2000);

Serial.println("Setup starting...");

Serial.println("I2C pin ...");
my_pins.addI2C(PinFunction::CODEC, SCLPIN, SDAPIN, ES8388ADDR, I2CSPEED, myWire);
Serial.println("I2S pin ...");
my_pins.addI2S(PinFunction::CODEC, MCLKPIN, BCLKPIN, WSPIN, DOPIN, DIPIN);


  /*Serial.println("Pins begin ..."); 
  my_pins.begin();

  Serial.println("Board begin ..."); 
  audio_board.begin();*/

Serial.println("I2S begin ...");
auto i2s_config = i2s_out_stream.defaultConfig(RXTX_MODE);
i2s_config.input_device = ADC_INPUT_ALL;
i2s_config.output_device = DAC_OUTPUT_LINE2;
i2s_config.sd_active = false;
//i2s_config.i2s.fmt = I2S_NORMAL;
//i2s_config.i2s.mode = MODE_SLAVE;
i2s_config.copyFrom(audio_info);
i2s_out_stream.begin(i2s_config); // this should apply I2C and I2S configuration

// Setup sine wave
Serial.println("Sine wave begin...");
sound_stream.begin(audio_info);
sine_wave.begin(audio_info, N_C6); // 493.88 Hz

Serial.println("Setup completed ...");
}

// Arduino loop - copy sound to out
void loop() { copier.copy(); }
```

**Reply by pschatzmann** [reply, marked as answer]
That was an interesting one! 
I changed to pass the pins by reference and now my proposed sketch from above is working.
Just update this library...

16:31:11.038 -> Setup starting...
16:31:11.038 -> I2C pin ...
16:31:11.038 -> I2S pin ...
16:31:11.038 -> I2S begin ...
16:31:11.038 -> Debug:   AudioBoard::pins::begin
16:31:11.038 -> Debug:   DriverPins::begin
16:31:11.046 -> Debug:   DriverPins::begin::setupPinMode
16:31:11.046 -> Debug:   DriverPins::begin::SPI
16:31:11.046 -> Debug:   DriverPins::begin::I2C
16:31:11.046 -> Debug:   DriverPins::begin::I2C port:32
16:31:11.046 -> Debug:   PinsI2C::begin for 7
16:31:11.046 -> Info:    setting up I2C scl: 32, sda: 33
16:31:11.047 -> Info:    Setting i2c clock: 100000
16:31:11.047 -> Debug:   AudioBoard::pins::begin::returned:true
16:31:11.047 -> Debug:   AudioBoard::driver::begin
16:31:11.047 -> Debug:   AudioDriver::begin:pins
16:31:11.055 -> Debug:   AudioDriver::begin:setSPI
16:31:11.055 -> Debug:   AudioDriver::begin:setConfig
16:31:11.055 -> Debug:   es8388_init
16:31:11.055 -> Debug:   i2c_bus_check: addr=0x10
16:31:11.055 -> Info:    Found ES8388
16:31:11.055 -> Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x19 value=0x4
16:31:11.055 -> Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x1 value=0x50

**Reply by lufa07** [reply]
Amazing solve Phil :) !

I do not understand the whole issue yet however, I hear now some tone on my earphones.

Excited to check this on my custom board now :)
